pkg/apply/prune/prune.go

```diff
@@ -99,6 +99,16 @@ func (p *Pruner) Prune(
 	for _, obj := range objs {
 		id := object.UnstructuredToObjMetadata(obj)
 		klog.V(5).Infof("evaluating prune filters (object: %q)", id)
+
+		// UID will change if the object is deleted and re-created.
+		uid := obj.GetUID()
+		if uid == "" {
+			err := object.NotFound([]interface{}{"metadata", "uid"}, "")
+			taskContext.SendEvent(eventFactory.CreateFailedEvent(id, err))
+			taskContext.InventoryManager().AddFailedDelete(id)
```

Comment:
When the object is deleted or re-created by 3rd party, we think that the prune operation is successful.

Reply:
If the object is deleted _before the applier runs_ then yes, you are correct.

However, if the object is deleted after the initial GET and before the DELETE, the current code would error with a 404 NotFound error.

And if the object were deleted and recreated after the initial GET and before the DELETE, the current code would delete the new object, thinking it was the old object, even if it didn't have the same inventory-id annotation.

This change I added causes the DELETE to error if the UID has changed at all since the GET was made, which fixes both of the above use cases by erroring with a clear error that the precondition was not satisfied because the UID has changed. This makes the behavior more like a server-side-apply, which errors if there's irreconcilable drift. The user can then simply retry the applier to re-do the GET & DELETE calls.

Reply:
This particular line just errors if the UID is empty, which should never happen, because we always pull it from the server at the beginning of the applier when deciding what to prune.

Reply:
The cases that you described sounds good to me.

For the case that the object is deleted by the applier and recreated the 3rd party, it is after or during the applier run.
In this case, the UID doesn't change from GET to DELETE, but will change during the Wait period.
Could you point me to where it it handle?

Reply:
This PR doesn't solve that bug yet. I just needed the UID to put into the `Inventory.ObjectStatus.UID`, so I figured I'd fix a delete bug I saw while i was at it. I'll make another PR to fix the bug we caught with the Config Sync e2e test.

Reply:
Now that the UID is in the TaskContext tho, it unblocks using it in the WaitTask.

Reply:
Makes sense. Thank you for explaining this.